File: ruby/121.best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  buy_price = prices[0]
  max_profit = 0

  prices.each do |price|
    if price < buy_price
      buy_price = price
    elsif price - buy_price > max_profit
      max_profit = price - buy_price
    end
  end

  max_profit
end

File: rust/src/121.best-time-to-buy-and-sell-stock.rs
use crate::Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        prices
            .iter()
            .fold((i32::MAX, 0), |(buy_price, max_profit), &price| {
                (buy_price.min(price), max_profit.max(price - buy_price))
            })
            .1
    }
}
